lib/Sema/ConstraintSystem.cpp

```diff
@@ -1359,24 +1359,24 @@ ConstraintSystem::getTypeOfMemberReference(
   // When accessing protocol members with an existential base, replace
   // the 'Self' type parameter with the existential type, since formally
   // the access will operate on existentials and not type parameters.
-  if (!isDynamicResult &&
-      baseObjTy->isExistentialType() &&
-      outerDC->getSelfProtocolDecl()) {
-    auto selfTy = replacements[
-      cast<GenericTypeParamType>(outerDC->getSelfInterfaceType()
-                                 ->getCanonicalType())];
-    type = type.transform([&](Type t) -> Type {
-      if (auto *selfTy = t->getAs<DynamicSelfType>())
-        t = selfTy->getSelfType();
-      if (t->is<TypeVariableType>())
-        if (t->isEqual(selfTy))
-          return baseObjTy;
-      if (auto *metatypeTy = t->getAs<MetatypeType>())
-        if (metatypeTy->getInstanceType()->isEqual(selfTy))
-          return ExistentialMetatypeType::get(baseObjTy);
-      return t;
-    });
-  }
+//  if (!isDynamicResult &&
```

Comment:
Was this causing problems?

Reply:
This piece of code was replacing `Self` with the protocol type, preventing the constraint system from treating them as archetypes and trying to resolve a nested type that can be resolved. So nothing like `resolveNestedType` or `simplifyType` was called and the compiler consistently tripped on an unhandled coercion every time there was such a nested type. I commented it out as a hack because I didn't know how else to convince the constraint system to pay attention. Here's where I need some help.

Reply:
@slavapestov I've been playing with this part to better understand the impact on the constraint system, and whether I can do something more reasonable than just commenting out existing code.
Having a better picture, I now don't really understand why we need this code at all. 

When the base type is existential, we open and bind it to a type variable as the `Self` generic parameter. In contrast, the archetype acts as `Self` in case of an archetype base.
https://github.com/AnthonyLatsis/swift/blob/master/lib/Sema/ConstraintSystem.cpp#L1351-1353
But then comes this code that replaces `Self` with the existential base when the base is existential. So says the comment, but actually, it replaces the type parameter that represents `Self`, which is bound to the opened existential base, which is an archetype. When we make this replacement, the resulting type becomes free of type variables and thus isn't simplified because `simplifyType` does not accept types without type variables. This is OK hitherto, since not a single existential base with associated types would have tripped over this code, and no associated types means there is nothing to simplify. The replacement in question probably slightly optimizes the CS and constraint generation, but still...
Anyways, existential bases that have concrete associated types are a different story, because these associated types must be simplified to their concrete equivalents. This does not happen without the type variables that continue getting replaced because `simplifyType` isn't called (as it would in case of an archetype *base*). So from Swift's perspective, all that has to be done to make it work is replace, say, `func foo(arg: Proto)` with `func foo<T: Proto>(arg: T)`.
___
I see two approaches to this issue, but the second is mentioned merely for the sake of completeness:
* Remove the code in question, or only resort to these type variable replacements when the base type doesn't have dependent member types (`openedArchetype->getAllNestedTypes().size() == 0` is this efficient though?). 
* Preserve replacements, but replace with the opened archetype instead so that the simplification can take place. Also, we would still need to allow types without types variables in `simplifyType` (not as trivial as removing an `if` statement).